packages/react-components/react-timepicker-compat-preview/src/components/TimePicker/useTimePicker.tsx

```diff
@@ -163,17 +163,24 @@ const useStableDateAnchor = (providedDate: Date | undefined, startHour: Hour, en
  * - TimePicker loses focus, signifying a possible change.
  */
 const useSelectTimeFromValue = (state: TimePickerState, callback: TimePickerProps['onTimeSelect']) => {
-  const { activeOption, freeform, validateFreeFormTime, options, submittedText, setActiveOption, value } = state;
+  const { activeOption, freeform, validateFreeFormTime, submittedText, setActiveOption, value } = state;
 
   // Base Combobox has activeOption default to first option in dropdown even if it doesn't match input value, and Enter key will select it.
   // This effect ensures that the activeOption is cleared when the input doesn't match any option.
   // This behavior is specific to a freeform TimePicker where the input value is treated as a valid time even if it's not in the dropdown.
-  const isValueOptionPrefix = value ? options.some(({ text }) => text.indexOf(value) === 0) : false;
   React.useEffect(() => {
-    if (freeform && value && !isValueOptionPrefix) {
-      setActiveOption(undefined);
-    }
-  }, [freeform, isValueOptionPrefix, setActiveOption, value]);
+    setActiveOption(prevActiveOption => {
+      if (!freeform) {
+        return prevActiveOption;
+      }
```

Comment:
nit: you can put the `freeform` check outside the call to `setActiveOption`:
```suggestion
    if (freeform) {
      setActiveOption(prevActiveOption => {
```

Reply:
You could also move the `value` check outside the call as well:
```ts
    if (freeform && value) {
      setActiveOption(prevActiveOption => {
        if (prevActiveOption?.text?.indexOf(value) === 0)
          return prevActiveOption;
        }
        return undefined;
      });
    }
```

Reply:
Yeah this looks cleaner :) updated.